File: atlas/examples/typescript/packages/infrastructure/src/catalog/InMemoryCatalogRepository.ts
import type { CatalogRepository } from '@atlas-example/application';
import type { CatalogItem } from '@atlas-example/domain';
import { orderBy } from 'lodash-es';

/**
 * Stores catalog items in memory while preserving deterministic query ordering.
 */
export class InMemoryCatalogRepository implements CatalogRepository {
  private readonly items = new Map<string, CatalogItem>();

  /**
   * Stores or replaces one item by identity.
   *
   * @param item - Valid domain item.
   */
  public save(item: CatalogItem): void {
    this.items.set(item.id, item);
  }

  /**
   * Lists stored items by normalized title and identifier.
   *
   * @returns Immutable deterministic item view.
   */
  public list(): readonly CatalogItem[] {
    return orderBy(
      [...this.items.values()],
      [(item) => item.title.toLocaleLowerCase(), (item) => item.id],
      ['asc', 'asc'],
    );
  }
}

File: atlas/examples/kotlin/infrastructure/src/main/kotlin/dev/atlas/example/infrastructure/catalog/InMemoryCatalogRepository.kt
package dev.atlas.example.infrastructure.catalog

import dev.atlas.example.application.catalog.CatalogRepository
import dev.atlas.example.domain.catalog.CatalogItem
import org.apache.commons.lang3.StringUtils

/**
 * Stores catalog items in memory while preserving deterministic query ordering.
 */
class InMemoryCatalogRepository : CatalogRepository {
    private val items = mutableMapOf<String, CatalogItem>()

    /**
     * Stores or replaces one item by identity.
     *
     * @param item Valid domain item.
     */
    override fun save(item: CatalogItem) {
        items[item.id] = item
    }

    /**
     * Lists stored items by normalized title and identifier.
     *
     * @return Immutable deterministic item view.
     */
    override fun list(): List<CatalogItem> {
        return items.values.sortedWith(
            compareBy<CatalogItem> { item -> StringUtils.lowerCase(item.title) }
                .thenBy { item -> item.id }
        )
    }
}
